news-bundle/contao/modules/ModuleNewsReader.php

```diff
@@ -159,6 +159,11 @@ protected function compile()
 			{
 				$htmlHeadBag->setMetaRobots($objArticle->robots);
 			}
+
+			if ($objArticle->canonicalLink)
+			{
+				$htmlHeadBag->setCanonicalUri($objArticle->canonicalLink);
+			}
```

Comment:
We can't directly set the canonical URL this way, can we? Isn't the following logic missing then?

https://github.com/contao/contao/blob/e0c479fd92e42f21cbb223d9d68fe80ce2193ba4/core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php#L88-L101

Reply:
You're right, but that logic is totally independent from pages and should always be applied when setting a canonical link. But the HtmlHeadBag can't do that, because it is not a service 🤔

Reply:
May be we need to apply this when _reading_ from the `HtmlHeadBag` then?

Reply:
As far as I remember @Toflar the HtmlHeadBag is supposed to be _raw_, so it should always contain a valid URL. But we have a million similar places in Contao where we previously apply service-logic in non-services, so maybe we should discuss about generic helpers or another solution to this?

Reply:
Bags can have access to services. Check the factory :)

Reply:
Fixed in 1b8f61c60519ac4c5226b353eef580aa1ec05087